### ISSUE TITLE
Rise 5.90 source code

### ISSUE THREAD

**Issue by gabrielvicenteYT** [Spinyfish/Client-Hall-Of-Fame#24]
Ok, first of all, we have to understand WHY I want the source code to be in the Client Hall of Fame. As I was checking Rise 5.90, I found that it skidded from LiquidBounce, here's the evidence:
LiquidBounce b72 silent rotation strafe code (Rotation.kt, rotation utils):
```
    /**
     * Apply strafe to player
     *
     * @author bestnub
     */
    fun applyStrafeToPlayer(event: StrafeEvent) {
        val player = MinecraftInstance.mc.thePlayer
        val dif = ((MathHelper.wrapAngleTo180_float(player.rotationYaw - this.yaw
                - 23.5f - 135)
                + 180) / 45).toInt()

        val yaw = this.yaw

        val strafe = event.strafe
        val forward = event.forward
        val friction = event.friction

        var calcForward = 0f
        var calcStrafe = 0f

        when (dif) {
            0 -> {
                calcForward = forward
                calcStrafe = strafe
            }
            1 -> {
                calcForward += forward
                calcStrafe -= forward
                calcForward += strafe
                calcStrafe += strafe
            }
            2 -> {
                calcForward = strafe
                calcStrafe = -forward
            }
            3 -> {
                calcForward -= forward
                calcStrafe -= forward
                calcForward += strafe
                calcStrafe -= strafe
            }
            4 -> {
                calcForward = -forward
                calcStrafe = -strafe
            }
            5 -> {
                calcForward -= forward
                calcStrafe += forward
                calcForward -= strafe
                calcStrafe -= strafe
            }
            6 -> {
                calcForward = -strafe
                calcStrafe = forward
            }
            7 -> {
                calcForward += forward
                calcStrafe += forward
                calcForward -= strafe
                calcStrafe += strafe
            }
        }

        if (calcForward > 1f || calcForward < 0.9f && calcForward > 0.3f || calcForward < -1f || calcForward > -0.9f && calcForward < -0.3f) {
            calcForward *= 0.5f
        }

        if (calcStrafe > 1f || calcStrafe < 0.9f && calcStrafe > 0.3f || calcStrafe < -1f || calcStrafe > -0.9f && calcStrafe < -0.3f) {
            calcStrafe *= 0.5f
        }

        var d = calcStrafe * calcStrafe + calcForward * calcForward

        if (d >= 1.0E-4f) {
            d = MathHelper.sqrt_float(d)
            if (d < 1.0f) d = 1.0f
            d = friction / d
            calcStrafe *= d
            calcForward *= d
            val yawSin = MathHelper.sin((yaw * Math.PI / 180f).toFloat())
            val yawCos = MathHelper.cos((yaw * Math.PI / 180f).toFloat())
            player.motionX += calcStrafe * yawCos - calcForward * yawSin.toDouble()
            player.motionZ += calcForward * yawCos + calcStrafe * yawSin.toDouble()
        }
    }
}
```
Rise 5.90 silent rotation strafe code (Scaffold):

```
    public static void silentRotationStrafe(final StrafeEvent event, final float yaw) {
        final int dif = (int) ((MathHelper.wrapAngleTo180_float(mc.thePlayer.rotationYaw - yaw - 23.5F - 135.0F) + 180.0F) / 45.0F);
        final float strafe = event.getStrafe();
        final float forward = event.getForward();
        final float friction = event.getFriction();
        float calcForward = 0.0F;
        float calcStrafe = 0.0F;
        switch (dif) {
            case 0: {
                calcForward = forward;
                calcStrafe = strafe;
                break;
            }

            case 1: {
                calcForward += forward;
                calcStrafe -= forward;
                calcForward += strafe;
                calcStrafe += strafe;
                break;
            }

            case 2: {
                calcForward = strafe;
                calcStrafe = -forward;
                break;
            }

            case 3: {
                calcForward -= forward;
                calcStrafe -= forward;
                calcForward += strafe;
                calcStrafe -= strafe;
                break;
            }

            case 4: {
                calcForward = -forward;
                calcStrafe = -strafe;
                break;
            }

            case 5: {
                calcForward -= forward;
                calcStrafe += forward;
                calcForward -= strafe;
                calcStrafe -= strafe;
                break;
            }

            case 6: {
                calcForward = -strafe;
                calcStrafe = forward;
                break;
            }

            case 7: {
                calcForward += forward;
                calcStrafe += forward;
                calcForward -= strafe;
                calcStrafe += strafe;
                break;
            }
        }

        if (calcForward > 1.0F || (calcForward < 0.9F && calcForward > 0.3F) || calcForward < -1.0F || (calcForward > -0.9F && calcForward < -0.3F))
            calcForward *= 0.5F;

        if (calcStrafe > 1.0F || (calcStrafe < 0.9F && calcStrafe > 0.3F) || calcStrafe < -1.0F || (calcStrafe > -0.9F && calcStrafe < -0.3F))
            calcStrafe *= 0.5F;

        float d;
        if ((d = calcStrafe * calcStrafe + calcForward * calcForward) >= 1.0E-4F) {
            if ((d = MathHelper.sqrt_float(d)) < 1.0F) {
                d = 1.0F;
            }
            d = friction / d;
            final float yawSin = MathHelper.sin((float) (yaw * Math.PI / 180.0));
            final float yawCos = MathHelper.cos((float) (yaw * Math.PI / 180.0));
            mc.thePlayer.motionX += (calcStrafe *= d) * yawCos - (calcForward *= d) * yawSin;
            mc.thePlayer.motionZ += calcForward * yawCos + calcStrafe * yawSin;
        }
    }
```

…you'd think that it would end there, right? But no, Rise devs INSIST on skidding from LiquidBounce, so here we go!
Rise 5.90 MoveUtils:
```
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 * Ty for borrow lol
 */
package dev.rise.util.player;

import dev.rise.event.impl.other.MoveEvent;
import lombok.experimental.UtilityClass;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.network.play.client.C03PacketPlayer;
import net.minecraft.potion.Potion;
import net.minecraft.util.MathHelper;
import net.minecraft.util.MovementInput;

import javax.vecmath.Vector2d;

/**
 * MoveUtils, used to move in modules
 *
 * @author Liquidbounce
 * @since 13/02/2021
 */

@UtilityClass
public final class MoveUtil {

    private final Minecraft mc = Minecraft.getMinecraft();

    /**
     * Used to get the players speed
     */
    public double getSpeed() {
        // nigga hypot heavy
        return Math.hypot(mc.thePlayer.motionX, mc.thePlayer.motionZ);
    }

    /**
     * Sets current speed to itself make strafe
     */
    public void strafe() {
        strafe(getSpeed());
    }

    /**
     * Checks if the player is moving
     */
    public boolean isMoving() {
        return mc.thePlayer != null && (mc.thePlayer.movementInput.moveForward != 0F || mc.thePlayer.movementInput.moveStrafe != 0F);
    }

    public void stop() {
        mc.thePlayer.motionX = mc.thePlayer.motionZ = 0;
    }

    /**
     * Sets players speed, with floats
     */
    public void strafe(final float speed) {
        if (!isMoving()) return;

        final double yaw = getDirection();

        mc.thePlayer.motionX = -MathHelper.sin((float) yaw) * speed;
        mc.thePlayer.motionZ = MathHelper.cos((float) yaw) * speed;
    }

    /**
     * Used to get the players speed, with doubles
     */
    public void strafe(final double speed) {
        if (!isMoving()) return;

        final double yaw = getDirection();
        mc.thePlayer.motionX = -MathHelper.sin((float) yaw) * speed;
        mc.thePlayer.motionZ = MathHelper.cos((float) yaw) * speed;
    }

    public void forward(final double speed) {
        final double yaw = getDirection();

        mc.thePlayer.motionX = -Math.sin(yaw) * speed;
        mc.thePlayer.motionZ = Math.cos(yaw) * speed;
    }

    public double moveSpeed() {
        if (mc.gameSettings.keyBindSprint.isKeyDown()) {
            if (mc.thePlayer.isPotionActive(Potion.moveSpeed)) {
                if (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).getAmplifier() + 1 == 1) {
                    return 0.18386012061481244;
                } else {
                    return 0.21450346015841276;
                }
            } else {
                return 0.15321676228437875;
            }
        } else {
            if (mc.thePlayer.isPotionActive(Potion.moveSpeed)) {
                if (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).getAmplifier() + 1 == 1) {
                    return 0.14143085686761;
                } else {
                    return 0.16500264553372018;
                }
            } else {
                return 0.11785905094607611;
            }
        }
    }

    /**
     * Used to get the players speed with a custom yaw
     */
    public void strafe(final double speed, float yaw) {
        if (!isMoving()) return;

        if (EntityPlayer.movementYaw != null) {
            yaw = EntityPlayer.movementYaw;
        }

        mc.thePlayer.motionX = -MathHelper.sin(yaw) * speed;
        mc.thePlayer.motionZ = MathHelper.cos(yaw) * speed;
    }

    /**
     * Gets the direction of were the player is looking
     */
    public double getDirection() {
        float rotationYaw = mc.thePlayer.rotationYaw;

        if (mc.thePlayer.moveForward < 0F) rotationYaw += 180F;

        float forward = 1F;

        if (mc.thePlayer.moveForward < 0F) forward = -0.5F;
        else if (mc.thePlayer.moveForward > 0F) forward = 0.5F;

        if (mc.thePlayer.moveStrafing > 0F) rotationYaw -= 90F * forward;
        if (mc.thePlayer.moveStrafing < 0F) rotationYaw += 90F * forward;

        return Math.toRadians(rotationYaw);
    }

    public double getDirectionWrappedTo90() {
        float rotationYaw = mc.thePlayer.rotationYaw;

        if (mc.thePlayer.moveForward < 0F && mc.thePlayer.moveStrafing == 0F) rotationYaw += 180F;

        final float forward = 1F;

        if (mc.thePlayer.moveStrafing > 0F) rotationYaw -= 90F * forward;
        if (mc.thePlayer.moveStrafing < 0F) rotationYaw += 90F * forward;

        return Math.toRadians(rotationYaw);
    }

    public double getDirection(final float yaw) {
        float rotationYaw = yaw;

        if (EntityPlayer.movementYaw != null) {
            rotationYaw = EntityPlayer.movementYaw;
        }

        if (mc.thePlayer.moveForward < 0F) rotationYaw += 180F;

        float forward = 1F;

        if (mc.thePlayer.moveForward < 0F) forward = -0.5F;
        else if (mc.thePlayer.moveForward > 0F) forward = 0.5F;

        if (mc.thePlayer.moveStrafing > 0F) rotationYaw -= 90F * forward;
        if (mc.thePlayer.moveStrafing < 0F) rotationYaw += 90F * forward;

        return Math.toRadians(rotationYaw);
    }

    /**
     * Used to get base movement speed
     */
    public double getBaseMoveSpeed() {
        double baseSpeed = 0.2873D;

        if (mc.thePlayer.isPotionActive(Potion.moveSpeed))
            baseSpeed *= 1.0D + 0.2D * (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).getAmplifier() + 1);

        return baseSpeed;
    }

    public double getBaseMoveSpeedOther() {
        double baseSpeed = 0.2875D;

        if (mc.thePlayer.isPotionActive(Potion.moveSpeed))
            baseSpeed *= 1.0D + 0.2D * (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).getAmplifier() + 1);

        return baseSpeed;
    }

    public double getJumpMotion(float motionY) {
        final Potion potion = Potion.jump;

        if (mc.thePlayer.isPotionActive(potion)) {
            final int amplifier = mc.thePlayer.getActivePotionEffect(potion).getAmplifier();
            motionY += (amplifier + 1) * 0.1F;
        }

        return motionY;
    }

    public double getPredictedMotionY(final double motionY) {
        return (motionY - 0.08) * 0.98F;
    }

    public void setMoveEventSpeed(final MoveEvent moveEvent, final double moveSpeed) {
        setMoveEvent(moveEvent, moveSpeed, mc.thePlayer.rotationYaw, mc.thePlayer.movementInput.moveStrafe, mc.thePlayer.movementInput.moveForward);
    }

    public boolean isOnGround(final double height) {
        return !mc.theWorld.getCollidingBoundingBoxes(mc.thePlayer, mc.thePlayer.getEntityBoundingBox().offset(0.0, -height, 0.0)).isEmpty();
    }

    public void sendMotion(final double speed, final double dist) {
        final Vector2d motion = getMotion(dist);
        double d = dist;
        while (d < speed) {
            mc.thePlayer.moveEntityNoEvent(motion.x, 0, motion.y);
            PacketUtil.sendPacket(new C03PacketPlayer.C04PacketPlayerPosition(mc.thePlayer.posX, mc.thePlayer.posY, mc.thePlayer.posZ, true));
            d += dist;
        }
    }

    public Vector2d getMotion(final double moveSpeed) {
        final MovementInput movementInput = mc.thePlayer.movementInput;

        double moveForward = movementInput.moveForward;
        double moveStrafe = movementInput.moveStrafe;

        double rotationYaw = mc.thePlayer.rotationYaw;
        if (moveForward != 0.0D || moveStrafe != 0.0D) {
            if (moveStrafe > 0) {
                moveStrafe = 1;
            } else if (moveStrafe < 0) {
                moveStrafe = -1;
            }
            if (moveForward != 0.0D) {
                if (moveStrafe > 0.0D) {
                    rotationYaw += moveForward > 0.0D ? -45 : 45;
                } else if (moveStrafe < 0.0D) {
                    rotationYaw += moveForward > 0.0D ? 45 : -45;
                }
                moveStrafe = 0.0D;
                if (moveForward > 0.0D) {
                    moveForward = 1.0D;
                } else if (moveForward < 0.0D) {
                    moveForward = -1.0D;
                }
            }
            rotationYaw *= 0.995;
            final double cos = Math.cos(Math.toRadians(rotationYaw + 90.0F));
            final double sin = Math.sin(Math.toRadians(rotationYaw + 90.0F));
            return new Vector2d(moveForward * moveSpeed * cos
                    + moveStrafe * moveSpeed * sin, moveForward * moveSpeed * sin
                    - moveStrafe * moveSpeed * cos);
        }
        return new Vector2d(0, 0);
    }

    public void setMoveEvent(final MoveEvent moveEvent, final double moveSpeed, final float pseudoYaw, final double pseudoStrafe, final double pseudoForward) {
        double forward = pseudoForward;
        double strafe = pseudoStrafe;
        float yaw = pseudoYaw;

        if (forward != 0.0D) {
            if (strafe > 0.0D) {
                yaw += ((forward > 0.0D) ? -45 : 45);
            } else if (strafe < 0.0D) {
                yaw += ((forward > 0.0D) ? 45 : -45);
            }
            strafe = 0.0D;
            if (forward > 0.0D) {
                forward = 1.0D;
            } else if (forward < 0.0D) {
                forward = -1.0D;
            }
        }
        if (strafe > 0.0D) {
            strafe = 1.0D;
        } else if (strafe < 0.0D) {
            strafe = -1.0D;
        }
        final double mx = Math.cos(Math.toRadians((yaw + 90.0F)));
        final double mz = Math.sin(Math.toRadians((yaw + 90.0F)));
        moveEvent.setX(forward * moveSpeed * mx + strafe * moveSpeed * mz);
        moveEvent.setZ(forward * moveSpeed * mz - strafe * moveSpeed * mx);
    }

}
```

This time, they even *admitted* to skidding / “borrowing”!
Now, they're going to fully credit LiquidBounce! FINALLY! (ClickUtil.java)
```
    /**
     * Gets a randomized click delay between 2 cps bounds.
     * Credits to LiquidBounce.
     *
     * @param minCPS The minimum cps allowed.
     * @param maxCPS The maximum cps allowed.
     * @return Click delay.
     */
    public long randomClickDelay(final int minCPS, final int maxCPS) {
        return (long) ((long) (((Math.random() * (1000 / minCPS - 1000 / maxCPS + 1)) + 1000 / maxCPS)) * Math.random());
    }
```

**Comment by gabrielvicenteYT** [comment]
TL;DR: Rise 5.90 skidded a ton of utils from LiquidBounce, and should be clowned upon.

**Comment by Spinyfish** [comment]
What the fish? Rise 5.90? Skidded? Oh frick... I wonder if the general code is bad too? 
Make a pull request I'll add it mwah 💋

**Comment by Spinyfish** [comment]
Thanks for bringing this to my attention! Please make a pull request and I will merge it 😄

**Comment by Briiqn** [comment]
> TL;DR: Rise 5.90 skidded a ton of utils from LiquidBounce, and should be clowned upon.

Wdym they've been doing it since 4.0, they even credited lb for their module manager

**Comment by Briiqn** [comment]
> Ok, first of all, we have to understand WHY I want the source code to be in the Client Hall of Fame. As I was checking Rise 5.90, I found that it skidded from LiquidBounce, here's the evidence:
> LiquidBounce b72 silent rotation strafe code (Rotation.kt, rotation utils):
> ```
>     /**
>      * Apply strafe to player
>      *
>      * @author bestnub
>      */
>     fun applyStrafeToPlayer(event: StrafeEvent) {
>         val player = MinecraftInstance.mc.thePlayer
>         val dif = ((MathHelper.wrapAngleTo180_float(player.rotationYaw - this.yaw
>                 - 23.5f - 135)
>                 + 180) / 45).toInt()
> 
>         val yaw = this.yaw
> 
>         val strafe = event.strafe
>         val forward = event.forward
>         val friction = event.friction
> 
>         var calcForward = 0f
>         var calcStrafe = 0f
> 
>         when (dif) {
>             0 -> {
>                 calcForward = forward
>                 calcStrafe = strafe
>             }
>             1 -> {
>                 calcForward += forward
>                 calcStrafe -= forward
>                 calcForward += strafe
>                 calcStrafe += strafe
>             }
>             2 -> {
>                 calcForward = strafe
>                 calcStrafe = -forward
>             }
>             3 -> {
>                 calcForward -= forward
>                 calcStrafe -= forward
>                 calcForward += strafe
>                 calcStrafe -= strafe
>             }
>             4 -> {
>                 calcForward = -forward
>                 calcStrafe = -strafe
>             }
>             5 -> {
>                 calcForward -= forward
>                 calcStrafe += forward
>                 calcForward -= strafe
>                 calcStrafe -= strafe
>             }
>             6 -> {
>                 calcForward = -strafe
>                 calcStrafe = forward
>             }
>             7 -> {
>                 calcForward += forward
>                 calcStrafe += forward
>                 calcForward -= strafe
>                 calcStrafe += strafe
>             }
>         }
> 
>         if (calcForward > 1f || calcForward < 0.9f && calcForward > 0.3f || calcForward < -1f || calcForward > -0.9f && calcForward < -0.3f) {
>             calcForward *= 0.5f
>         }
> 
>         if (calcStrafe > 1f || calcStrafe < 0.9f && calcStrafe > 0.3f || calcStrafe < -1f || calcStrafe > -0.9f && calcStrafe < -0.3f) {
>             calcStrafe *= 0.5f
>         }
> 
>         var d = calcStrafe * calcStrafe + calcForward * calcForward
> 
>         if (d >= 1.0E-4f) {
>             d = MathHelper.sqrt_float(d)
>             if (d < 1.0f) d = 1.0f
>             d = friction / d
>             calcStrafe *= d
>             calcForward *= d
>             val yawSin = MathHelper.sin((yaw * Math.PI / 180f).toFloat())
>             val yawCos = MathHelper.cos((yaw * Math.PI / 180f).toFloat())
>             player.motionX += calcStrafe * yawCos - calcForward * yawSin.toDouble()
>             player.motionZ += calcForward * yawCos + calcStrafe * yawSin.toDouble()
>         }
>     }
> }
> ```
> Rise 5.90 silent rotation strafe code (Scaffold):
> 
> ```
>     public static void silentRotationStrafe(final StrafeEvent event, final float yaw) {
>         final int dif = (int) ((MathHelper.wrapAngleTo180_float(mc.thePlayer.rotationYaw - yaw - 23.5F - 135.0F) + 180.0F) / 45.0F);
>         final float strafe = event.getStrafe();
>         final float forward = event.getForward();
>         final float friction = event.getFriction();
>         float calcForward = 0.0F;
>         float calcStrafe = 0.0F;
>         switch (dif) {
>             case 0: {
>                 calcForward = forward;
>                 calcStrafe = strafe;
>                 break;
>             }
> 
>             case 1: {
>                 calcForward += forward;
>                 calcStrafe -= forward;
>                 calcForward += strafe;
>                 calcStrafe += strafe;
>                 break;
>             }
> 
>             case 2: {
>                 calcForward = strafe;
>                 calcStrafe = -forward;
>                 break;
>             }
> 
>             case 3: {
>                 calcForward -= forward;
>                 calcStrafe -= forward;
>                 calcForward += strafe;
>                 calcStrafe -= strafe;
>                 break;
>             }
> 
>             case 4: {
>                 calcForward = -forward;
>                 calcStrafe = -strafe;
>                 break;
>             }
> 
>             case 5: {
>                 calcForward -= forward;
>                 calcStrafe += forward;
>                 calcForward -= strafe;
>                 calcStrafe -= strafe;
>                 break;
>             }
> 
>             case 6: {
>                 calcForward = -strafe;
>                 calcStrafe = forward;
>                 break;
>             }
> 
>             case 7: {
>                 calcForward += forward;
>                 calcStrafe += forward;
>                 calcForward -= strafe;
>                 calcStrafe += strafe;
>                 break;
>             }
>         }
> 
>         if (calcForward > 1.0F || (calcForward < 0.9F && calcForward > 0.3F) || calcForward < -1.0F || (calcForward > -0.9F && calcForward < -0.3F))
>             calcForward *= 0.5F;
> 
>         if (calcStrafe > 1.0F || (calcStrafe < 0.9F && calcStrafe > 0.3F) || calcStrafe < -1.0F || (calcStrafe > -0.9F && calcStrafe < -0.3F))
>             calcStrafe *= 0.5F;
> 
>         float d;
>         if ((d = calcStrafe * calcStrafe + calcForward * calcForward) >= 1.0E-4F) {
>             if ((d = MathHelper.sqrt_float(d)) < 1.0F) {
>                 d = 1.0F;
>             }
>             d = friction / d;
>             final float yawSin = MathHelper.sin((float) (yaw * Math.PI / 180.0));
>             final float yawCos = MathHelper.cos((float) (yaw * Math.PI / 180.0));
>             mc.thePlayer.motionX += (calcStrafe *= d) * yawCos - (calcForward *= d) * yawSin;
>             mc.thePlayer.motionZ += calcForward * yawCos + calcStrafe * yawSin;
>         }
>     }
> ```
> 
> …you'd think that it would end there, right? But no, Rise devs INSIST on skidding from LiquidBounce, so here we go!
> Rise 5.90 MoveUtils:
> ```
> /*
>  * LiquidBounce Hacked Client
>  * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
>  * https://github.com/CCBlueX/LiquidBounce/
>  * Ty for borrow lol
>  */
> package dev.rise.util.player;
> 
> import dev.rise.event.impl.other.MoveEvent;
> import lombok.experimental.UtilityClass;
> import net.minecraft.client.Minecraft;
> import net.minecraft.entity.player.EntityPlayer;
> import net.minecraft.network.play.client.C03PacketPlayer;
> import net.minecraft.potion.Potion;
> import net.minecraft.util.MathHelper;
> import net.minecraft.util.MovementInput;
> 
> import javax.vecmath.Vector2d;
> 
> /**
>  * MoveUtils, used to move in modules
>  *
>  * @author Liquidbounce
>  * @since 13/02/2021
>  */
> 
> @UtilityClass
> public final class MoveUtil {
> 
>     private final Minecraft mc = Minecraft.getMinecraft();
> 
>     /**
>      * Used to get the players speed
>      */
>     public double getSpeed() {
>         // nigga hypot heavy
>         return Math.hypot(mc.thePlayer.motionX, mc.thePlayer.motionZ);
>     }
> 
>     /**
>      * Sets current speed to itself make strafe
>      */
>     public void strafe() {
>         strafe(getSpeed());
>     }
> 
>     /**
>      * Checks if the player is moving
>      */
>     public boolean isMoving() {
>         return mc.thePlayer != null && (mc.thePlayer.movementInput.moveForward != 0F || mc.thePlayer.movementInput.moveStrafe != 0F);
>     }
> 
>     public void stop() {
>         mc.thePlayer.motionX = mc.thePlayer.motionZ = 0;
>     }
> 
>     /**
>      * Sets players speed, with floats
>      */
>     public void strafe(final float speed) {
>         if (!isMoving()) return;
> 
>         final double yaw = getDirection();
> 
>         mc.thePlayer.motionX = -MathHelper.sin((float) yaw) * speed;
>         mc.thePlayer.motionZ = MathHelper.cos((float) yaw) * speed;
>     }
> 
>     /**
>      * Used to get the players speed, with doubles
>      */
>     public void strafe(final double speed) {
>         if (!isMoving()) return;
> 
>         final double yaw = getDirection();
>         mc.thePlayer.motionX = -MathHelper.sin((float) yaw) * speed;
>         mc.thePlayer.motionZ = MathHelper.cos((float) yaw) * speed;
>     }
> 
>     public void forward(final double speed) {
>         final double yaw = getDirection();
> 
>         mc.thePlayer.motionX = -Math.sin(yaw) * speed;
>         mc.thePlayer.motionZ = Math.cos(yaw) * speed;
>     }
> 
>     public double moveSpeed() {
>         if (mc.gameSettings.keyBindSprint.isKeyDown()) {
>             if (mc.thePlayer.isPotionActive(Potion.moveSpeed)) {
>                 if (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).getAmplifier() + 1 == 1) {
>                     return 0.18386012061481244;
>                 } else {
>                     return 0.21450346015841276;
>                 }
>             } else {
>                 return 0.15321676228437875;
>             }
>         } else {
>             if (mc.thePlayer.isPotionActive(Potion.moveSpeed)) {
>                 if (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).getAmplifier() + 1 == 1) {
>                     return 0.14143085686761;
>                 } else {
>                     return 0.16500264553372018;
>                 }
>             } else {
>                 return 0.11785905094607611;
>             }
>         }
>     }
> 
>     /**
>      * Used to get the players speed with a custom yaw
>      */
>     public void strafe(final double speed, float yaw) {
>         if (!isMoving()) return;
> 
>         if (EntityPlayer.movementYaw != null) {
>             yaw = EntityPlayer.movementYaw;
>         }
> 
>         mc.thePlayer.motionX = -MathHelper.sin(yaw) * speed;
>         mc.thePlayer.motionZ = MathHelper.cos(yaw) * speed;
>     }
> 
>     /**
>      * Gets the direction of were the player is looking
>      */
>     public double getDirection() {
>         float rotationYaw = mc.thePlayer.rotationYaw;
> 
>         if (mc.thePlayer.moveForward < 0F) rotationYaw += 180F;
> 
>         float forward = 1F;
> 
>         if (mc.thePlayer.moveForward < 0F) forward = -0.5F;
>         else if (mc.thePlayer.moveForward > 0F) forward = 0.5F;
> 
>         if (mc.thePlayer.moveStrafing > 0F) rotationYaw -= 90F * forward;
>         if (mc.thePlayer.moveStrafing < 0F) rotationYaw += 90F * forward;
> 
>         return Math.toRadians(rotationYaw);
>     }
> 
>     public double getDirectionWrappedTo90() {
>         float rotationYaw = mc.thePlayer.rotationYaw;
> 
>         if (mc.thePlayer.moveForward < 0F && mc.thePlayer.moveStrafing == 0F) rotationYaw += 180F;
> 
>         final float forward = 1F;
> 
>         if (mc.thePlayer.moveStrafing > 0F) rotationYaw -= 90F * forward;
>         if (mc.thePlayer.moveStrafing < 0F) rotationYaw += 90F * forward;
> 
>         return Math.toRadians(rotationYaw);
>     }
> 
>     public double getDirection(final float yaw) {
>         float rotationYaw = yaw;
> 
>         if (EntityPlayer.movementYaw != null) {
>             rotationYaw = EntityPlayer.movementYaw;
>         }
> 
>         if (mc.thePlayer.moveForward < 0F) rotationYaw += 180F;
> 
>         float forward = 1F;
> 
>         if (mc.thePlayer.moveForward < 0F) forward = -0.5F;
>         else if (mc.thePlayer.moveForward > 0F) forward = 0.5F;
> 
>         if (mc.thePlayer.moveStrafing > 0F) rotationYaw -= 90F * forward;
>         if (mc.thePlayer.moveStrafing < 0F) rotationYaw += 90F * forward;
> 
>         return Math.toRadians(rotationYaw);
>     }
> 
>     /**
>      * Used to get base movement speed
>      */
>     public double getBaseMoveSpeed() {
>         double baseSpeed = 0.2873D;
> 
>         if (mc.thePlayer.isPotionActive(Potion.moveSpeed))
>             baseSpeed *= 1.0D + 0.2D * (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).getAmplifier() + 1);
> 
>         return baseSpeed;
>     }
> 
>     public double getBaseMoveSpeedOther() {
>         double baseSpeed = 0.2875D;
> 
>         if (mc.thePlayer.isPotionActive(Potion.moveSpeed))
>             baseSpeed *= 1.0D + 0.2D * (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).getAmplifier() + 1);
> 
>         return baseSpeed;
>     }
> 
>     public double getJumpMotion(float motionY) {
>         final Potion potion = Potion.jump;
> 
>         if (mc.thePlayer.isPotionActive(potion)) {
>             final int amplifier = mc.thePlayer.getActivePotionEffect(potion).getAmplifier();
>             motionY += (amplifier + 1) * 0.1F;
>         }
> 
>         return motionY;
>     }
> 
>     public double getPredictedMotionY(final double motionY) {
>         return (motionY - 0.08) * 0.98F;
>     }
> 
>     public void setMoveEventSpeed(final MoveEvent moveEvent, final double moveSpeed) {
>         setMoveEvent(moveEvent, moveSpeed, mc.thePlayer.rotationYaw, mc.thePlayer.movementInput.moveStrafe, mc.thePlayer.movementInput.moveForward);
>     }
> 
>     public boolean isOnGround(final double height) {
>         return !mc.theWorld.getCollidingBoundingBoxes(mc.thePlayer, mc.thePlayer.getEntityBoundingBox().offset(0.0, -height, 0.0)).isEmpty();
>     }
> 
>     public void sendMotion(final double speed, final double dist) {
>         final Vector2d motion = getMotion(dist);
>         double d = dist;
>         while (d < speed) {
>             mc.thePlayer.moveEntityNoEvent(motion.x, 0, motion.y);
>             PacketUtil.sendPacket(new C03PacketPlayer.C04PacketPlayerPosition(mc.thePlayer.posX, mc.thePlayer.posY, mc.thePlayer.posZ, true));
>             d += dist;
>         }
>     }
> 
>     public Vector2d getMotion(final double moveSpeed) {
>         final MovementInput movementInput = mc.thePlayer.movementInput;
> 
>         double moveForward = movementInput.moveForward;
>         double moveStrafe = movementInput.moveStrafe;
> 
>         double rotationYaw = mc.thePlayer.rotationYaw;
>         if (moveForward != 0.0D || moveStrafe != 0.0D) {
>             if (moveStrafe > 0) {
>                 moveStrafe = 1;
>             } else if (moveStrafe < 0) {
>                 moveStrafe = -1;
>             }
>             if (moveForward != 0.0D) {
>                 if (moveStrafe > 0.0D) {
>                     rotationYaw += moveForward > 0.0D ? -45 : 45;
>                 } else if (moveStrafe < 0.0D) {
>                     rotationYaw += moveForward > 0.0D ? 45 : -45;
>                 }
>                 moveStrafe = 0.0D;
>                 if (moveForward > 0.0D) {
>                     moveForward = 1.0D;
>                 } else if (moveForward < 0.0D) {
>                     moveForward = -1.0D;
>                 }
>             }
>             rotationYaw *= 0.995;
>             final double cos = Math.cos(Math.toRadians(rotationYaw + 90.0F));
>             final double sin = Math.sin(Math.toRadians(rotationYaw + 90.0F));
>             return new Vector2d(moveForward * moveSpeed * cos
>                     + moveStrafe * moveSpeed * sin, moveForward * moveSpeed * sin
>                     - moveStrafe * moveSpeed * cos);
>         }
>         return new Vector2d(0, 0);
>     }
> 
>     public void setMoveEvent(final MoveEvent moveEvent, final double moveSpeed, final float pseudoYaw, final double pseudoStrafe, final double pseudoForward) {
>         double forward = pseudoForward;
>         double strafe = pseudoStrafe;
>         float yaw = pseudoYaw;
> 
>         if (forward != 0.0D) {
>             if (strafe > 0.0D) {
>                 yaw += ((forward > 0.0D) ? -45 : 45);
>             } else if (strafe < 0.0D) {
>                 yaw += ((forward > 0.0D) ? 45 : -45);
>             }
>             strafe = 0.0D;
>             if (forward > 0.0D) {
>                 forward = 1.0D;
>             } else if (forward < 0.0D) {
>                 forward = -1.0D;
>             }
>         }
>         if (strafe > 0.0D) {
>             strafe = 1.0D;
>         } else if (strafe < 0.0D) {
>             strafe = -1.0D;
>         }
>         final double mx = Math.cos(Math.toRadians((yaw + 90.0F)));
>         final double mz = Math.sin(Math.toRadians((yaw + 90.0F)));
>         moveEvent.setX(forward * moveSpeed * mx + strafe * moveSpeed * mz);
>         moveEvent.setZ(forward * moveSpeed * mz - strafe * moveSpeed * mx);
>     }
> 
> }
> ```
> 
> This time, they even *admitted* to skidding / “borrowing”!
> Now, they're going to fully credit LiquidBounce! FINALLY! (ClickUtil.java)
> ```
>     /**
>      * Gets a randomized click delay between 2 cps bounds.
>      * Credits to LiquidBounce.
>      *
>      * @param minCPS The minimum cps allowed.
>      * @param maxCPS The maximum cps allowed.
>      * @return Click delay.
>      */
>     public long randomClickDelay(final int minCPS, final int maxCPS) {
>         return (long) ((long) (((Math.random() * (1000 / minCPS - 1000 / maxCPS + 1)) + 1000 / maxCPS)) * Math.random());
>     }
> ```

Also the clickutil credit is funny because anyone with half a brain could make randomised clickdelay (ms)

**Comment by Briiqn** [comment]
Yet does not properly follow the gplv3 licensing lb and only "credits" as in the compiled version does not contain comments nor is there any mention of any use of lb code at all that the average end user at the time would notice (such as the mainmenu credits)

**Comment by Briiqn** [comment]
Either way the client is dead and nobody noticed it for over a year now because they had so much influence in the com that nobody would even think that they would skid

**Comment by gabrielvicenteYT** [comment]
> Either way the client is dead and nobody noticed it for over a year now because they had so much influence in the com that nobody would even think that they would skid

Yeah, serious conditioning. I'd like it if people just stuck with LiquidBounce or made their own hacked client, or a fork of an already good one (paid or free, just give credits ffs)
